Add 'genkey' verb to ukify

keszybz opened the pull request.

poettering commented on src/ukify/ukify.py:
```diff
@@ -729,6 +744,108 @@ def make_uki(opts):
+    from cryptography import x509
+    import cryptography.hazmat.primitives as hp
+
+    key = hp.asymmetric.rsa.generate_private_key(
```
there should be a comment here that explains the choice of RSA here and the key size, i.e. the fact that UEFI SecureBoot in the wild only really supports RSA up to some key length (2k?)

keszybz replied:
I just copied what our docs and/or mkosi was doing. I'd be happy to add something, but I don't really know too enough to write the text myself.

poettering replied:
msft says rsa 2k:

https://learn.microsoft.com/en-us/windows-hardware/manufacture/desktop/windows-secure-boot-key-creation-and-management-guidance?view=windows-11#12-public-key-cryptography

maybe just add a link to that as a comment

keszybz replied:
Added now.

poettering commented on src/ukify/ukify.py:
```diff
@@ -729,6 +744,108 @@ def make_uki(opts):
+    for priv_key, pub_key, _ in key_path_groups(opts):
+        priv_key_pem, pub_key_pem = generate_priv_pub_key_pair()
+
+        print(f'Writing private key to {priv_key}')
```
why not call these PCR private key? to make clear this one is about PCR signing rather than SecureBoot signing?

keszybz replied:
"PCR private key" is unclear.  I changed the texts to "Wring private key for PCR signing to …".  But suggestions for better wording are welcome.

I made the change locally, I'll push if there are other comments.

poettering replied:
"private key for PCR signing" sounds ok to me

poettering commented on man/ukify.xml:
```diff
@@ -498,6 +499,36 @@ Phases=enter-initrd:leave-initrd
       <literal>debug</literal> with SBAT metadata referring to the owner of the addon.</para>
     </example>
+
+    <example>
```
hmmm, so you add the example here, but never document the "genkey" verb in the first place...

poettering commented:
btw, can we make python-cryptography work with openssl engines? sooner or later we really need to hook this up to pkcs11 or some other HSM thing, so that people can sign PCR policies with keys on their HSM...

keszybz commented:
Hmm, what is the `reviewed/needs-rework` label for? Did you maybe forget to post some comments?

DaanDeMeyer commented:
If cryptography doesn't support specifying openssl engines, shouldn't we shell out to openssl instead of using cryptography so that we can easily add engine support in the future?

behrmann commented on src/ukify/ukify.py:
```diff
@@ -25,24 +25,30 @@
                     IO,
                     Optional,
+                    Sequence,
+                    Tuple,
```
Starting with 3.9 you can use `tuple` for typing where `Tuple` would have been necessary before.

keszybz replied:
I always get lost between all the different versions we support in different places. But indeed, ukify has an explicit check for 3.9, so `tuple` can be used.

behrmann commented on src/ukify/ukify.py:
```diff
@@ -345,6 +360,17 @@ def check_inputs(opts):
+        (opts.sb_key, opts.sb_cert),
+        *((priv_key, pub_key)
+          for priv_key, pub_key, _ in key_path_groups(opts)))
+    for path in paths:
+        if path and path.exists():
+            raise ValueError(f'{path} is present')
```
```suggestion
    if any(path and path.exists() for path in paths):
        raise ValueError(f'{path} is present')
```

keszybz replied:
No, that doesn't work, because `path` is not defined. I think it's important to include the path in the error message.

behrmann commented on src/ukify/ukify.py:
```diff
@@ -370,6 +396,19 @@ def combine_signatures(pcrsigs):
+    if not opts.pcr_private_keys:
+        return
+
+    n_priv = len(opts.pcr_private_keys or ())
```
```suggestion
    n_priv = len(opts.pcr_private_keys)
```
since if `opts.pcr_private_keys` is falsey, it will return before this line

poettering commented:
> for the missing man page changes

(I mean, as long as the man page addition is not forgotten i don't care if it's added in this pr or another)

esposem commented:
There are a lot of conflicts between this PR and mine (https://github.com/systemd/systemd/pull/27621). Since it seems that your PR go in wayy faster than outsiders, I would suggest to first merge this one and then I will adapt mine. Otherwise it is just additional unnecessary work to do.

keszybz commented:
> If cryptography doesn't support specifying openssl engines, shouldn't we shell out to openssl instead of using cryptography so that we can easily add engine support in the future?

My thinking was that the `genkey` verb will be used by people doing local signing. IIUC, openssl engine support would be for more "serious" use cases, where the key would be pregenerated using some other process.

keszybz commented:
Updated following the comments above. I added a new commit to actually document the `genkey` verb. The text addition is small, the diff is expanded by the need to insert a subsection layer.

poettering commented:
lgtm, but I'll leave the final review to someone with better python credentials than me

keszybz commented:
@behrmann please take another look.

behrmann commented on src/ukify/ukify.py:
```diff
@@ -711,6 +743,112 @@ def make_uki(opts):
+    ).issuer_name(
+        x509.Name([x509.NameAttribute(x509.oid.NameOID.COMMON_NAME, common_name)])
+    ).not_valid_before(
+        datetime.datetime.today() - ONE_DAY
```
Why not just
```suggestion
        datetime.datetime.now()
```
or better
```suggestion
        datetime.datetime.utcnow()
```
to avoid timezone ambiguities.

It probably doesn't matter much, but from glancing over it I wondered about the time it's already valid before existing.

keszybz replied:
I copied that code from the cryptography docs. But this really doesn't make any sense. I'll change it like suggested.

behrmann commented on src/ukify/ukify.py:
```diff
@@ -711,6 +743,112 @@ def make_uki(opts):
+    ).not_valid_before(
+        datetime.datetime.today() - ONE_DAY
+    ).not_valid_after(
+        datetime.datetime.today() + ONE_DAY * valid_days
```
Probably not an issue, but if you want the exact validity period: `datetime.datetime.today` ticks with `time.time`, so there will be some nanosecond difference between the two `datetime.datetime.today` invocations.

keszybz replied:
I changed this to use a local variable to avoid the double lookup.

behrmann commented on src/ukify/ukify.py:
```diff
@@ -711,6 +743,112 @@ def make_uki(opts):
+def generate_key_cert_pair(
+        common_name: str,
+        keylength: int = 2048,
+        valid_days: int = 365 * 10,  # TODO: can we drop the expiration date?
```
`cryptography` seems to to want one and error out with a `ValueError` if one is not given - I don't know if the standard requires one. Also not knowing the standard, could it be an issue that ten years in the future will soon be after 19 January 2038? 

Anyway, the key validity would probably be something that people might want to set as an option and in mkosi we have an option for that. (We also have one for the keylength, but in light of the below comment that's probably unnecessary)

keszybz replied:
Done now.

behrmann reviewed (commented):
Python-wise LGTM, only wondered about some some certificate things I don't know enough about. 

I think `cryptography` is a good library choice, but I'm sure some people will be unhappy about it, since at least part of the primitives are implemented in Rust nowadays. In mkosi we just shell out to `openssl`. I don't think this matters much for an optional thing like this, with plenty of alternatives  to generate the keys.

behrmann commented on src/ukify/ukify.py:
```diff
@@ -711,6 +743,116 @@ def make_uki(opts):
+
+    key = hp.asymmetric.rsa.generate_private_key(
+        public_exponent=65537,
+        key_size=keylength)
```
```suggestion
        key_size=keylength,
    )
```
to match the style of the below lines.

keszybz commented:
I'll take it as a yes.

keszybz commented:
jammy-ppc64el: TEST-70-TPM2:                       FAIL     (145 s)
mkosi / ci (centos, 8): something fails, but I'm not sure what. "Boot Centos Qemu". I'll restart it.

keszybz commented:
mkosi/ci (centos, 8) seems to time out:
`[  281.121998] systemd[1]: systemd-hwdb-update.service: start operation timed out. Terminating.`
I guess that's not related, even though it happened in the second run too.

keszybz commented:
Now all the CIs passed, except those two. Let's merge.
